### PR TITLE
add geolocation docs

### PR DESCRIPTION
**Pull request by patrickkettner**
fixes GoogleChromeLabs/Extension-Docs#24

Changes proposed in this pull request:
document how to use geolocation in various extension surfaces

### CODE REVIEW

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+```
+
+#### Using geolocation in popup or sidepanel
+
+If you want to use geolocation within a [popup](popup) or [side panel](sidepanel), it is very straightforward. Popups are just web documents, and therefore have access to the normal DOM APIs. You can access `navigator.geolocation` directly. The only difference from standard web sites would be that you need to use the manifest.json `permission` field to request the `geolocation` permission.  If you do not include the permission, you _will_ still have access to `navigator.geolocation`. However, any attempt to use it will cause an immediate error, the same as if the user rejected the request. You can see a sample of this [here](popup-sample).
+
+
+#### Using geolocation in a content script
+
+Just like a popup, a [content script](content)  has full access to the DOM API, therefore you will go through the normal user prompting permission flow. That means that adding  `geolocation` to your `permissions` _will not_ automatically give you access to the end users' geolocation information. You can see a sample of this [here](contentscript-sample)
```
Comment (sebastianbenz) [inline site/en/docs/extensions/mv3/geolocation/index.md:195]: "Just like a popup, a [content script](content) has full access to the DOM API, therefore you will go through the normal user prompting permission flow." This sounds like popup and content script behave the same, but according to the rest of the paragraph it doesn't.  Maybe rephrase.

Does `world` set to `ISOLATED` vs `main` make a difference to the behavior?

**Review by sebastianbenz — CHANGES_REQUESTED** [review]

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+title: "geolocation"
+seoTitle: "How to use geolocation in Chrome Extensions"
+date: 2023-06-20
+description: How to use geolocation in Chrome Extensions service worker, popup, sidepanel or content script
+---
+
+<!-- Copy and paste the converted output. -->
+
+
+## How to use geolocation in Chrome Extensions
```
[the remark this answers is not in the document]
Comment (jpmedley) [reply site/en/docs/extensions/mv3/geolocation/index.md:12]: +100

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+```js
+chrome.offscreen.createDocument({
+  url: 'offscreen.html',
+  reasons: [chrome.offscreen.Reason.GEOLOCATION || chrome.offscreen.Reason.DOM_SCRAPING],
+  justification: 'geolocation access',
+});
+```
+
+Note that when you access an offscreen document, you need to include a `reason`. The `geolocation` reason was not originally available, so we specify a fallback of `DOM\_SCRAPING`, and explain in our `justification` section what the code is actually doing. This information is used by the Chrome Web Store's review process to ensure offscreen documents are being used for a valid purpose.
```
[the remark this answers is not in the document]
Comment (jpmedley) [reply site/en/docs/extensions/mv3/geolocation/index.md:126]: I wouldn't.

File: site/en/docs/extensions/mv3/geolocation/index.md (hunk 1 above)
Comment (patrickkettner) [reply site/en/docs/extensions/mv3/geolocation/index.md:197]: good catch - a however transformed into a therefore.

changing the world does not impact the permission flow, is that what you are asking?

Comment (sebastianbenz) [reply site/en/docs/extensions/mv3/geolocation/index.md]: yes

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        justification: 'add justification for geolocation use here',
+      });
+
+      await creating;
+      creating = null;
+    }
+  }
+}
+
+async function getGeolocation() {
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply site/en/docs/extensions/mv3/geolocation/index.md]: Can you please also address this: " Might also be worth quickly mentioning why we need hasDocument and setupOffscreendoc (this can be short with a link to the offscreen docs)."

[the remark this answers is not in the document]
Comment (patrickkettner) [reply site/en/docs/extensions/mv3/geolocation/index.md]: sorry, meant to ask - there are brief comments there currently, and I do not believe I can link to content from within a comment - are you wanting me to add an inline reference to `https://developer.chrome.com/docs/extensions/reference/offscreen/#example` or add an additional section before/after the sample that explains those steps?

[the remark this answers is not in the document]
Comment (sebastianbenz) [reply site/en/docs/extensions/mv3/geolocation/index.md]: One approach could be to add an aside before the code snippet saying something like:

"A Chrome extension can only have one open offscreen document at a time and offscreen documents should be closed once they are no longer needed. To learn more about how to create and close offscreen documents, checkout the offscreen documentation."

**Review by sebastianbenz — APPROVED** [review]
Looks good from a technical POV once the offscreen doc link is added.

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    ...
+  "permissions": [
+    ...
+   "offscreen"
+  ],
+  ...
+}
+```
+
+After adding the `offscreen` permission, add an html file to our extension that will include your offscreen document. In this case, we are not using any of the content of the page, and so this can be a nearly blank file, it just needs to be a smal, html file that loads in our script.
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:42]:
```suggestion
After adding the `offscreen` permission, add an html file to your extension that includes your offscreen document. This case isn't using any of the content of the page, and so this can be a nearly blank file. It just needs to be a small, html file that loads in your script.
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+{% Label %}offscreen.html:{% endLabel %}
+
+```html
+<!doctype html>
+<title>offscreenDocument</title>
+<script src="offscreen.js"></script>
+```
+
+This file can be saved in the root of our project, as `offscreen.html`.
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:52]:
```suggestion
Save this file in the root of our project, as `offscreen.html`.
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+As mentioned we are including a script called [offscreen.js`, this will be another file we need to bundle with our extension. It will be where we are actually getting geolocation information, and sending that information back to our service worker. We can send and receive messages between it and our service worker.
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:54]:
```suggestion
As mentioned you need a script called `offscreen.js`. You'll also need to bundle this with your extension. It will be where you actually get geolocation information, and send that information back to the service worker. You can send and receive messages between it and your service worker.
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+  if (message.type !== 'get-geolocation') {
+    console.warn(`Unexpected message type received: '${message.type}'.`);
+    return;
+  }
+
+  // we can directly respond to the message from out service_worker with the
+  // provided `sendResponse` callback. But in order to be able to send an async
+  // response, we need to explictly return `true` in our onMessage handler
+  // As a result, we can't use async/await here - we'd implictly return a Promise.
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:74]:
```suggestion
  // You can directly respond to the message from the service worker with the
  // provided `sendResponse()` callback. But in order to be able to send an async
  // response, you need to explictly return `true` in the onMessage handler.
  // As a result, you can't use async/await here. You'd implictly return a Promise.
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+  getLocation().then((loc) => sendResponse(loc));
+
+  return true;
+}
+
+// getCurrentPosition returns a prototype based object, so the properties
+// end up being stripped off when sent over to our service worker. To get
+// around this, we deeply clone it
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:82]:
```suggestion
// getCurrentPosition() returns a prototype based object, so the properties
// end up being stripped off when sent to the service worker. To get
// around this, clone it deeply.
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+function clone(obj) {
+  const copy = {};
+  // Return the value of any non true object (typeof(null) is "object") directly.
+  // null will throw an error if you try to for/in it. We can just return
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:86]:
```suggestion
  // null will throw an error if you try to for/in it. Just return
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+    for (const p in obj) {
+      copy[p] = clone(obj[p]);
+    }
+  }
+  return copy;
+}
+
+async function getLocation() {
+  // we use a raw Promise here so we can pass `resolve` and `reject` into the
+  // callbacks for getCurrentPosition
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:100]:
```suggestion
  // Use a raw Promise here so you can pass `resolve` and `reject` into the
  // callbacks for getCurrentPosition().
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+      (loc) => resolve(clone(loc)),
+      // in case the user doesnt have/is blocking `geolocation`
+      (err) => reject(err)
+    );
+  });
+}
+```
+
+
+With that in place, we are now ready to access the Offscreen Document in our service worker!
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:112]:
```suggestion
With that in place, you are now ready to access the Offscreen Document in the service worker.
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+  reasons: [chrome.offscreen.Reason.GEOLOCATION || chrome.offscreen.Reason.DOM_SCRAPING],
+  justification: 'geolocation access',
+});
+```
+
+Note that when you access an offscreen document, you need to include a `reason`. The `geolocation` reason was not originally available, so we specify a fallback of `DOM\_SCRAPING`, and explain in our `justification` section what the code is actually doing. This information is used by the Chrome Web Store's review process to ensure offscreen documents are being used for a valid purpose.
+
+Once you have a reference to the Offscreen Document, you can send it a message to ask for it to give you updated geolocation information.
+
+{% Label %}service\_worker.js:{% endLabel %}
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:126]:
```suggestion
{% Label %}service_worker.js:{% endLabel %}
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+```js
+chrome.offscreen.createDocument({
+  url: 'offscreen.html',
+  reasons: [chrome.offscreen.Reason.GEOLOCATION || chrome.offscreen.Reason.DOM_SCRAPING],
+  justification: 'geolocation access',
+});
+```
+
+Note that when you access an offscreen document, you need to include a `reason`. The `geolocation` reason was not originally available, so we specify a fallback of `DOM\_SCRAPING`, and explain in our `justification` section what the code is actually doing. This information is used by the Chrome Web Store's review process to ensure offscreen documents are being used for a valid purpose.
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:122]:
```suggestion
Note that when you access an offscreen document, you need to include a `reason`. The `geolocation` reason was not originally available, so specify a fallback of `DOM_SCRAPING`, and explain in the `justification` section what the code is actually doing. This information is used by the Chrome Web Store's review process to ensure offscreen documents are being used for a valid purpose.
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+async function closeOffscreenDocument() {
+  if (!(await hasDocument())) {
+    return;
+  }
+  await chrome.offscreen.closeDocument();
+}
+```
+
+So now whenever you want to get the geolocation from your service worker, you just need to call
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:180]:
```suggestion
So now whenever you want to get the geolocation from your service worker, you just need to call:
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+```js
+const location = await getGeolocation()
+```
+
+### Using geolocation in popup or sidepanel
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:186]:
```suggestion
### Use geolocation in popup or side panel
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+
+If you want to use geolocation within a [popup](popup) or [side panel](sidepanel), it is very straightforward. Popups and sidepanels are just web documents, and therefore have access to the normal DOM APIs. You can access `navigator.geolocation` directly. The only difference from standard web sites would be that you need to use the manifest.json `permission` field to request the `geolocation` permission. If you do not include the permission, you _will_ still have access to `navigator.geolocation`. However, any attempt to use it will cause an immediate error, the same as if the user rejected the request. You can see a sample of this [here](popup-sample).
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:188]:
```suggestion
If you want to use geolocation within a [popup](popup) or [side panel](sidepanel), it is very straightforward. Popups and side panels are just web documents, and therefore have access to the normal DOM APIs. You can access `navigator.geolocation` directly. The only difference from standard web sites would be that you need to use the `manifest.json` `"permission"` field to request the `geolocation` permission. If you do not include the permission, you _will_ still have access to `navigator.geolocation`. However, any attempt to use it will cause an immediate error, the same as if the user rejected the request. You can see this in the [popup sample](popup-sample).
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+
+
+### Using geolocation in a content script
+
+Just like a popup, a [content script](content) has full access to the DOM API, however you will go through the normal user prompting permission flow. That means that adding `geolocation` to your `permissions` _will not_ automatically give you access to the end users' geolocation information. You can see a sample of this [here](contentscript-sample)
```
Comment (jpmedley) [inline site/en/docs/extensions/mv3/geolocation/index.md:193]:
```suggestion
Just like a popup, a [content script](content) has full access to the DOM API; however you will go through the normal user permission flow. That means that adding `geolocation` to your `permissions` _will not_ automatically give you access to the users' geolocation information. You can see this in the [content script sample](contentscript-sample)
```

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 19]
```diff
@@ -0,0 +1,204 @@
+---
+layout: "layouts/doc-post.njk"
+title: "Using geolocation"
+seoTitle: "How to use geolocation in Chrome Extensions"
+date: 2023-06-20
+description: How to use geolocation in Chrome Extensions service worker, popup, sidepanel or content script
+---
+
+If you want to get geolocation information in your Chrome extension, use the same [`navigator.geolocation`](geolocation) DOM API that any web site normally would. The reason this article exists is because Chrome extensions handle permission to sensitive data differently than normal websites. Geolocation is very sensitive data, and as such browsers take effort to make sure that people using browsers are fully aware and in control of when and where their exact location is shared.
```
[the remark this answers is not in the document]
Comment (patrickkettner) [reply site/en/docs/extensions/mv3/geolocation/index.md]: Appreciated! for some reason I was under the impression we should use person centered language, but on checking I guess that is more in the [a11y specific area](https://support.google.com/dspa-customer-care/answer/10506622?hl=en). ill make sure to be more terse in the future.

File: site/en/docs/extensions/mv3/geolocation/index.md [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+
+
+{% Img src="image/DXqUldooyJOUnj3qXSYLHbUgUI93/nYRDFWbb280yEp90XDTI.png", alt="A screenshot of the permission prompt you see when a website requests access to the geolocation api", width="800", height="419" %}
+
+Permissions are not the only difference. As mentioned above, `navigator.geolocation` is a \_\_DOM\_\_ API, that is, something that is a part of the APIs that make up web sites. As a result, it is not accessible inside Worker contexts, like the [extension service worker](sw) that is the backbone of manifest v3 extensions. You can absolutely still use `geolocation`, though. There are just nuances with how and where you use it.
+
+
+### Using geolocation in Service Workers
+
+There is no `navigator` object inside of service workers. It is only available inside of contexts that have access to a page's `document` object. To get access inside of a service worker, use an [`Offscreen Document`](offscreen). This is a feature of extensions that gives you access to an html file you can bundle with our extension. You can read more about Offscreen Documents in the [documentation](offscreen).
```
[the remark this answers is not in the document]
Comment (patrickkettner) [reply site/en/docs/extensions/mv3/geolocation/index.md:26]: removed

**Review by jpmedley — APPROVED** [review]
